simulator/src/testbench.js

```diff
@@ -1,3 +1,230 @@
+    for(let dataIO of data.groups[0].inputs.concat(data.groups[0].outputs)){
+        const matchIO = scope.Input.concat(scope.Output).some(function(simulatorIO) {
+            // TODO: label to identifier
+            return simulatorIO.label === dataIO.label && simulatorIO.bitWidth === dataIO.bitWidth;
+        });
+
+        if(!matchIO) return { status: "fail", message: "Some inputs/outputs not present in circuit or wrong bitwidths" };
+    }
+
+    // Validate presence of reset if test is sequential
+    if(data.type === "seq"){
+        const resetPresent = scope.Input.some(function(simulatorReset) { return simulatorReset.label === "RST" });
+
+        if(!(resetPresent)) return {status: "fail", message: "Reset(RST) not present in circuit"};
+    }
+
+    return { status: "ok" };
+}
+
+/**
+ * Checks if all the labels in the test are unique. Called by validate()
+ * @param {Object} data - Object containing Test Data
+ */
+function checkDistinctIdentifiers(data) {
+    const input_identifers = data.groups[0].inputs.map(function(input) { return input.label });
+    const output_identifiers = data.groups[0].outputs.map(function(output) { return output.label });
+    const identifiers = input_identifers.concat(output_identifiers);
+
+    return (new Set(identifiers)).size === identifiers.length;
+}
+
+/**
+ * Returns object of scope inputs and outputs keyed by their labels
+ * @param {Scope=} scope - the circuit
+ */
+function bindIO(data, scope) {
+    let inputs = {};
+    let outputs = {};
+    let clock, reset;
+    for(const dataInput of data.groups[0].inputs){
+        inputs[dataInput.label] = scope.Input.find(function(simulatorInput) {
+            return simulatorInput.label === dataInput.label
+        });
+    }
+
+    for(const dataOutput of data.groups[0].outputs){
+        outputs[dataOutput.label] = scope.Output.find(function(simulatorOutput) {
+            return simulatorOutput.label === dataOutput.label
+        });
+    }
+
+    if(data.type === "seq"){
+        clock = scope.Clock.find(function(simulatorOutput) {
+            return simulatorOutput.label === "CLK"
+        });
+
+        reset = scope.Input.find(function(simulatorOutput) {
+            return simulatorOutput.label === "RST"
+        });
+    }
+
+    return { inputs: inputs, outputs: outputs, clock: clock, reset: reset };
+}
+
+/**
+ * Ticks clock recursively one full cycle (Only used in testbench context)
+ * @param {Scope} scope - the circuit whose clock to be ticked
+ */
+function tickClock(scope){
+    scope.clockTick();
+    play();
+    scope.clockTick();
+    play();
+}
+
+/**
+ * Triggers reset (Only used in testbench context)
+ * @param {Input} reset - reset pin to be triggered
+ */
+function triggerReset(reset){
+    reset.state = 1;
+    play();
+    reset.state = 0;
+    play();
+}
+
+/**
+ * Creates deep clone of scope given to the function
+ * @param {Scope} scope - the scope to be cloned
+ */
+function cloneScope(scope) {
```

Comment:
I think cloning scope is not a good idea. But let us see.

Reply:
Yes, I dropped the idea, not using clone scope anymore